concept/type_/annotation.rs

```diff
@@ -82,11 +105,118 @@ impl AnnotationRegex {
         Self { regex }
     }
 
+    pub fn default() -> Self {
+        Self::new(".*".to_owned())
+    }
+
     pub fn regex(&self) -> &str {
         &self.regex
     }
 }
 
+impl Annotation {
+    pub fn category(&self) -> AnnotationCategory {
+        match self {
+            Self::Abstract(_) => AnnotationCategory::Abstract,
+            Self::Distinct(_) => AnnotationCategory::Distinct,
+            Self::Independent(_) => AnnotationCategory::Independent,
+            Self::Unique(_) => AnnotationCategory::Unique,
+            Self::Key(_) => AnnotationCategory::Key,
+            Self::Cardinality(_) => AnnotationCategory::Cardinality,
+            Self::Regex(_) => AnnotationCategory::Regex,
+            Self::Cascade(_) => AnnotationCategory::Cascade,
+        }
+    }
+}
+
+#[derive(Debug, Default, Copy, Clone, Eq, PartialEq, Hash)]
+pub struct AnnotationCascade;
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum AnnotationCategory {
+    Abstract,
+    Distinct,
+    Independent,
+    Unique,
+    Key,
+    Cardinality,
+    Regex,
+    Cascade,
+    // TODO: Subkey
+    // TODO: Values
+    // TODO: Range
+    // TODO: Replace
+}
+
+impl AnnotationCategory {
+    pub fn to_default_annotation(&self) -> Annotation {
+        // TODO: fn const for regex default
+        match self {
+            AnnotationCategory::Abstract => Annotation::Abstract(AnnotationAbstract),
+            AnnotationCategory::Distinct => Annotation::Distinct(AnnotationDistinct),
+            AnnotationCategory::Independent => Annotation::Independent(AnnotationIndependent),
+            AnnotationCategory::Unique => Annotation::Unique(AnnotationUnique),
+            AnnotationCategory::Key => Annotation::Key(AnnotationKey),
+            AnnotationCategory::Cardinality => Annotation::Cardinality(AnnotationCardinality::default()),
+            AnnotationCategory::Regex => Annotation::Regex(AnnotationRegex::default()),
+            AnnotationCategory::Cascade => Annotation::Cascade(AnnotationCascade),
+        }
+    }
+
+    pub fn compatible_to_declare_together(&self, other: AnnotationCategory) -> bool {
+        match self {
+            AnnotationCategory::Unique => match other {
+                AnnotationCategory::Key => false,
+                _ => true,
+            },
+            AnnotationCategory::Cardinality => match other {
+                AnnotationCategory::Key => false,
+                _ => true,
+            },
+            AnnotationCategory::Key => match other {
+                | AnnotationCategory::Unique | AnnotationCategory::Cardinality => false,
+                _ => true,
+            },
+            | AnnotationCategory::Abstract
+            | AnnotationCategory::Distinct
+            | AnnotationCategory::Independent
+            | AnnotationCategory::Regex
+            | AnnotationCategory::Cascade => true,
+        }
+    }
+
+    pub fn compatible_to_transitively_add(&self, other_to_add: AnnotationCategory) -> bool {
```

Comment:
A more relaxed check for compatibility on different levels. Used in two cases:
* When I want to set a new annotation and need to check if it is compatible with already transitively added annotations (if we already have `Key`, we can't set `Unique` or `Cardinality` on subtypes).
* When we collect stored transitive annotations for a type starting from its declared annotations (we already have declared `Key` on a subtype, so we don't need to take `Unique` or `Cardinality` from a supertype). We consider that all the stored annotations are correct (thanks to `compatible_to_declare_together`).

I didn't come up with a better name to specify that `&self` should be a part of a collection of already processed annotations where want to add `other_to_add` to. Feel free to suggest!